prow/statusreconciler/controller.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statusreconciler
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/maintenance/migratestatus/migrator"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/errorutil"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/plugins/trigger"
+)
+
+// NewController constructs a new controller to reconcile stauses on config change
+func NewController(continueOnError bool, kubeClient *kube.Client, githubClient *github.Client, configAgent *config.Agent, pluginAgent *plugins.ConfigAgent) *Controller {
+	return &Controller{
+		continueOnError: continueOnError,
+		kubeClient:      kubeClient,
+		githubClient:    githubClient,
+		configAgent:     configAgent,
+		pluginAgent:     pluginAgent,
+	}
+}
+
+// Controller reconciles statuses on PRs when config changes impact blocking presubmits
+type Controller struct {
+	continueOnError bool
+	kubeClient      *kube.Client
```

Comment:
These should all be interfaces so you can unit test, not concrete implementations.